File: debate-engine/src/orchestrator/argumentRegistry.ts
import type { Argument, FactCheck, Rebuttal } from "../types/index.js";

interface TrackedArgument {
  argument: Argument;
  raisedInRound: number;
  raisedBy: "advocate" | "critic";
  rebuttals: Rebuttal[];
  factChecks: FactCheck[];
}

export class ArgumentRegistry {
  private arguments: Map<string, TrackedArgument> = new Map();

  register(arg: Argument, round: number, agent: "advocate" | "critic"): void {
    this.arguments.set(arg.id, {
      argument: { ...arg },
      raisedInRound: round,
      raisedBy: agent,
      rebuttals: [],
      factChecks: [],
    });
  }

  addRebuttal(targetId: string, rebuttal: Rebuttal): void {
    const tracked = this.arguments.get(targetId);
    if (tracked) {
      tracked.rebuttals.push(rebuttal);
    }
  }

  addFactCheck(targetId: string, check: FactCheck): void {
    const tracked = this.arguments.get(targetId);
    if (tracked) {
      tracked.factChecks.push(check);
      if (check.verdict === "flawed") {
        tracked.argument.status = "rebutted";
      }
    }
  }

  updateStatus(argId: string, status: Argument["status"]): void {
    const tracked = this.arguments.get(argId);
    if (tracked) {
      tracked.argument.status = status;
    }
  }

  getActiveArguments(): Argument[] {
    return Array.from(this.arguments.values())
      .filter((t) => t.argument.status === "active" || t.argument.status === "modified")
      .map((t) => t.argument);
  }

  getActiveByAgent(agent: "advocate" | "critic"): Argument[] {
    return Array.from(this.arguments.values())
      .filter(
        (t) =>
          t.raisedBy === agent &&
          (t.argument.status === "active" || t.argument.status === "modified")
      )
      .map((t) => t.argument);
  }

  getAll(): TrackedArgument[] {
    return Array.from(this.arguments.values());
  }

  getSurvivorStats(): { total: number; survived: number; rebutted: number; conceded: number } {
    const all = Array.from(this.arguments.values());
    return {
      total: all.length,
      survived: all.filter(
        (t) => t.argument.status === "active" || t.argument.status === "modified"
      ).length,
      rebutted: all.filter((t) => t.argument.status === "rebutted").length,
      conceded: all.filter((t) => t.argument.status === "conceded").length,
    };
  }

  getChallengeCount(argId: string): number {
    const tracked = this.arguments.get(argId);
    if (!tracked) return 0;
    return tracked.rebuttals.length + tracked.factChecks.filter((fc) => fc.verdict === "flawed").length;
  }
}

File: maverickj/core/argument_registry.py
from maverickj.schemas.arguments import (
    Argument,
    ArgumentRecord,
    ArgumentStatus,
    FactCheck,
    FactCheckVerdict,
    Rebuttal,
)


class ArgumentRegistry:
    def __init__(self, data: dict[str, ArgumentRecord] | None = None):
        self._arguments: dict[str, ArgumentRecord] = dict(data) if data else {}

    def register(self, arg: Argument, round_num: int, agent: str) -> None:
        """Register a new argument."""
        self._arguments[arg.id] = ArgumentRecord(
            argument=arg,
            raised_in_round=round_num,
            raised_by=agent,
        )

    def update_status(self, arg_id: str, new_status: ArgumentStatus, reason: str = "") -> None:
        """Update the status of an argument."""
        if arg_id in self._arguments:
            record = self._arguments[arg_id]
            record.argument.status = new_status
            if reason:
                record.modification_history.append(reason)

    def add_rebuttal(self, arg_id: str, rebuttal: Rebuttal) -> None:
        """Add a rebuttal record to an argument."""
        if arg_id in self._arguments:
            self._arguments[arg_id].rebuttals.append(rebuttal)

    def add_fact_check(self, arg_id: str, check: FactCheck) -> None:
        """Add a fact-check record to an argument."""
        if arg_id in self._arguments:
            self._arguments[arg_id].fact_checks.append(check)
            if check.verdict == FactCheckVerdict.FLAWED:
                self.update_status(
                    arg_id,
                    ArgumentStatus.REBUTTED,
                    f"Fact-check: {check.explanation}",
                )

    def get_active_arguments(self, side: str | None = None) -> list[ArgumentRecord]:
        """Return all surviving arguments, optionally filtered by side."""
        results = [
            r for r in self._arguments.values()
            if r.argument.status in (ArgumentStatus.ACTIVE, ArgumentStatus.MODIFIED)
        ]
        if side:
            results = [r for r in results if r.raised_by == side]
        return results

    def get_survivor_stats(self) -> dict:
        """Return argument survival statistics."""
        total = len(self._arguments)
        active = len([
            r for r in self._arguments.values()
            if r.argument.status in (ArgumentStatus.ACTIVE, ArgumentStatus.MODIFIED)
        ])
        return {
            "total_raised": total,
            "survived": active,
            "rebutted": len([
                r for r in self._arguments.values()
                if r.argument.status == ArgumentStatus.REBUTTED
            ]),
            "conceded": len([
                r for r in self._arguments.values()
                if r.argument.status == ArgumentStatus.CONCEDED
            ]),
        }

    def to_dict(self) -> dict[str, ArgumentRecord]:
        return dict(self._arguments)
